### PR TITLE
add layer norm and expand size layers

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

. Implement Layer normalization
. Implement ExpandSize and TableOperationExpand that handling two tensors with different size
. Add corresponding python api

## How was this patch tested?

unit tests

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ExpandSize.scala [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+
+/**
+ * Expand tensor to configured size
+ * @param tgtSizes target tensor sizes, dim whose size is -1 will be ignored
+ * @tparam T Numeric type of parameter(e.g. weight, bias). Only support float/double now.
+ */
+class ExpandSize[T: ClassTag](tgtSizes: Array[Int])
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ExpandSize.scala]: preferable `targetSizes` ?

Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ExpandSize.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ExpandSize.scala [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    }
+    gradInput = _gradOutput
+    gradInput
+  }
+
+  override def toString: String = s"ExpandSize"
+}
+
+object ExpandSize {
+  def apply[@specialized(Float, Double) T: ClassTag](tgtSizes: Array[Int])
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ExpandSize.scala]: targetSize, please keep consistent with class definition

Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ExpandSize.scala]: ok, changed

**Comment by zhangxiaoli73** [comment]
jenkins pass: http://mlt-bddl01.sh.intel.com:8080/view/PR/job/BigDL-PRVN-lite/593/

File: pyspark/bigdl/nn/layer.py [hunk 3]
```diff
@@ -929,6 +929,40 @@ class FeedForwardNetwork(Layer):
...
+    Applies layer normalization.
+
+    >>> norm = LayerNormalization(8)
+    creating: createLayerNormalization
+    '''
+
+    def __init__(self, hidden_size, bigdl_type="float"):
+        super(LayerNormalization, self).__init__(None, bigdl_type, hidden_size)
+
+class TableOperationExpand(Layer):
```
Comment (wzhongyuan) [inline pyspark/bigdl/nn/layer.py]: `Expand ` is a verb

Comment (wzhongyuan) [reply pyspark/bigdl/nn/layer.py]: Probably just remove it

File: pyspark/bigdl/nn/layer.py [hunk 4]
```diff
@@ -929,6 +929,40 @@ class FeedForwardNetwork(Layer):
...
+    '''
+    When two tensors have different size, firstly expand small size tensor to large size tensor,
+    and then do table operation.
+
+    >>> norm = TableOperationExpand(CMulTable())
+    creating: createCMulTable
+    creating: createTableOperationExpand
+    '''
+
+    def __init__(self, operation_layer, expand_pos=2, bigdl_type="float"):
```
Comment (wzhongyuan) [inline pyspark/bigdl/nn/layer.py]: why we should specify the expand_pos ? can we infer from input ? it's determined by the input stead of layer itself right ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/LayerNormalization.scala [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+/**
+ * Implement x * weight vector + bias vector
+ * @param hiddenSize
+ * @param ev$1
+ * @param ev
+ * @tparam T The numeric type in this module parameters
+ */
+private[nn] class LayerLinear[T: ClassTag](val hiddenSize: Int)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/LayerNormalization.scala]: What's the difference with `Linear` ?

File: pyspark/bigdl/nn/layer.py (hunk 4 above)
Comment (zhangxiaoli73) [reply pyspark/bigdl/nn/layer.py]: yeah, good idea. Refine code and infer position from input

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/LayerNormalization.scala (hunk 5 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/LayerNormalization.scala]: This layer implements `input vector * weight vector + bias vector`. It should be an element-wise multiply operation.
If input is a tensor with two, then we should select tensors along its's first dim to do multiply.

So you see it is different with nn linear. Maybe name is not clear. How about `VectorLinear` or any other?

File: pyspark/bigdl/nn/layer.py (hunk 3 above)
Comment (zhangxiaoli73) [reply pyspark/bigdl/nn/layer.py]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/LayerNormalization.scala (hunk 5 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/LayerNormalization.scala]: `VectorProduct` ?

Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/LayerNormalization.scala]: ok, changed

**Comment by zhangxiaoli73** [comment]
jenkins pass: http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/597/

**Pull request merged**